File: app/src/main/java/com/krystianwsul/checkme/firebase/RemoteCustomTime.java
package com.krystianwsul.checkme.firebase;

import android.support.annotation.NonNull;
import android.support.v4.util.Pair;

import com.krystianwsul.checkme.domainmodel.CustomTime;
import com.krystianwsul.checkme.domainmodel.DomainFactory;
import com.krystianwsul.checkme.firebase.records.RemoteCustomTimeRecord;
import com.krystianwsul.checkme.utils.CustomTimeKey;
import com.krystianwsul.checkme.utils.time.DayOfWeek;
import com.krystianwsul.checkme.utils.time.HourMinute;
import com.krystianwsul.checkme.utils.time.TimePair;

import java.util.TreeMap;

class RemoteCustomTime implements CustomTime {
    @NonNull
    private final DomainFactory mDomainFactory;

    @NonNull
    private final RemoteProject mRemoteProject;

    @NonNull
    private final RemoteCustomTimeRecord mRemoteCustomTimeRecord;

    RemoteCustomTime(@NonNull DomainFactory domainFactory, @NonNull RemoteProject remoteProject, @NonNull RemoteCustomTimeRecord remoteCustomTimeRecord) {
        mDomainFactory = domainFactory;
        mRemoteProject = remoteProject;
        mRemoteCustomTimeRecord = remoteCustomTimeRecord;
    }

    @NonNull
    public String getId() {
        return mRemoteCustomTimeRecord.getId();
    }

    @NonNull
    @Override
    public String getName() {
        return mRemoteCustomTimeRecord.getName();
    }

    @NonNull
    @Override
    public HourMinute getHourMinute(@NonNull DayOfWeek dayOfWeek) {
        switch (dayOfWeek) {
            case SUNDAY:
                return new HourMinute(mRemoteCustomTimeRecord.getSundayHour(), mRemoteCustomTimeRecord.getSundayMinute());
            case MONDAY:
                return new HourMinute(mRemoteCustomTimeRecord.getMondayHour(), mRemoteCustomTimeRecord.getMondayMinute());
            case TUESDAY:
                return new HourMinute(mRemoteCustomTimeRecord.getTuesdayHour(), mRemoteCustomTimeRecord.getTuesdayMinute());
            case WEDNESDAY:
                return new HourMinute(mRemoteCustomTimeRecord.getWednesdayHour(), mRemoteCustomTimeRecord.getWednesdayMinute());
            case THURSDAY:
                return new HourMinute(mRemoteCustomTimeRecord.getThursdayHour(), mRemoteCustomTimeRecord.getThursdayMinute());
            case FRIDAY:
                return new HourMinute(mRemoteCustomTimeRecord.getFridayHour(), mRemoteCustomTimeRecord.getFridayMinute());
            case SATURDAY:
                return new HourMinute(mRemoteCustomTimeRecord.getSaturdayHour(), mRemoteCustomTimeRecord.getSaturdayMinute());
            default:
                throw new IllegalArgumentException("invalid day: " + dayOfWeek);
        }
    }

    @NonNull
    @Override
    public TreeMap<DayOfWeek, HourMinute> getHourMinutes() {
        TreeMap<DayOfWeek, HourMinute> hourMinutes = new TreeMap<>();
        for (DayOfWeek dayOfWeek : DayOfWeek.values())
            hourMinutes.put(dayOfWeek, getHourMinute(dayOfWeek));
        return hourMinutes;
    }

    @NonNull
    public String toString() {
        return getName();
    }

    @NonNull
    @Override
    public Pair<CustomTime, HourMinute> getPair() {
        return new Pair<>(this, null);
    }

    @NonNull
    @Override
    public TimePair getTimePair() {
        return new TimePair(new CustomTimeKey(mRemoteProject.getId(), mRemoteCustomTimeRecord.getId()), null);
    }

    @NonNull
    @Override
    public CustomTimeKey getCustomTimeKey() {
        return mDomainFactory.getCustomTimeKey(mRemoteProject.getId(), getId());
    }
}

File: app/src/main/java/com/krystianwsul/checkme/firebase/RemoteCustomTime.kt
package com.krystianwsul.checkme.firebase

import com.krystianwsul.checkme.firebase.records.RemoteCustomTimeRecord
import com.krystianwsul.common.domain.CustomTime
import com.krystianwsul.common.time.DayOfWeek
import com.krystianwsul.common.time.HourMinute
import com.krystianwsul.common.time.TimePair
import com.krystianwsul.common.utils.RemoteCustomTimeId
import java.util.*

abstract class RemoteCustomTime<T : RemoteCustomTimeId> : CustomTime {

    protected abstract val remoteProject: RemoteProject<T>

    abstract val remoteCustomTimeRecord: RemoteCustomTimeRecord<T>

    protected abstract val allRecords: List<RemoteCustomTimeRecord<*>>

    abstract val id: T

    override var name
        get() = remoteCustomTimeRecord.name
        set(value) {
            allRecords.forEach { it.name = value }
        }

    override val hourMinutes
        get() = DayOfWeek.values()
                .map { it to getHourMinute(it) }
                .toMap(TreeMap())

    override val timePair by lazy { TimePair(customTimeKey, null) }// possibly should get local key from DomainFactory (instead I have to do it in RemoteInstance)

    val projectId by lazy { remoteProject.id }

    override fun getHourMinute(dayOfWeek: DayOfWeek) = when (dayOfWeek) {
        DayOfWeek.SUNDAY -> HourMinute(remoteCustomTimeRecord.sundayHour, remoteCustomTimeRecord.sundayMinute)
        DayOfWeek.MONDAY -> HourMinute(remoteCustomTimeRecord.mondayHour, remoteCustomTimeRecord.mondayMinute)
        DayOfWeek.TUESDAY -> HourMinute(remoteCustomTimeRecord.tuesdayHour, remoteCustomTimeRecord.tuesdayMinute)
        DayOfWeek.WEDNESDAY -> HourMinute(remoteCustomTimeRecord.wednesdayHour, remoteCustomTimeRecord.wednesdayMinute)
        DayOfWeek.THURSDAY -> HourMinute(remoteCustomTimeRecord.thursdayHour, remoteCustomTimeRecord.thursdayMinute)
        DayOfWeek.FRIDAY -> HourMinute(remoteCustomTimeRecord.fridayHour, remoteCustomTimeRecord.fridayMinute)
        DayOfWeek.SATURDAY -> HourMinute(remoteCustomTimeRecord.saturdayHour, remoteCustomTimeRecord.saturdayMinute)
    }

    override fun toString() = name

    abstract fun delete()

    override fun setHourMinute(dayOfWeek: DayOfWeek, hourMinute: HourMinute) = allRecords.forEach { it.setHourMinute(dayOfWeek, hourMinute) }
}
